### ISSUE TITLE
Prepayment failing with Failed prepayment: UnknownPaymentHash causes problems

### ISSUE THREAD

**Issue by OpinionatedGeek** [lightninglabs/loop#54]
I tried to do a Loop Out:
```
2019-05-11 12:48:08 loop out --channel 632050960802447360 --amt 500000
```

Here's what appeared in the Loop logs:
```
2019-05-11 12:48:08.872 [INF] CLIENT: Offchain swap destination: 03fb2a0ca79c005f493f1faa83071d3a937cf220d4051dc48b8fe3a087879cf14a
2019-05-11 12:48:12.485 [INF] LOOPD: Loop out request received
2019-05-11 12:48:12.489 [INF] CLIENT: LoopOut 0.005 BTC to bc1qyryv9rnvf7p994xjc69dlyp2vlsdxez8yv2rdh (channel: 0xc00050b4d0)
2019-05-11 12:48:12.494 [INF] CLIENT: Initiating swap request at height 575535
2019-05-11 12:48:12.742 [INF] CLIENT: 8f60fe Htlc address: bc1qdph5mvhpsnra73mnjzvvayhxzuzf8a5f9tzy0eggcz2judrel54qgl2xhm
2019-05-11 12:48:12.751 [INF] CLIENT: 8f60fe state Initiated
2019-05-11 12:48:12.751 [INF] CLIENT: 8f60fe Sending swap payment lnbc4857070n...
2019-05-11 12:48:12.751 [INF] CLIENT: 8f60fe Sending prepayment lnbc279140n1...
2019-05-11 12:48:12.752 [INF] CLIENT: 8f60fe Register conf ntfn for swap script on chain (hh=575535)
2019-05-11 12:48:12.757 [INF] CLIENT: 8f60fe Checking preimage reveal height 575565 exceeded (height 575535)
2019-05-11 12:48:12.757 [INF] CLIENT: 8f60fe Waiting for either htlc on-chain confirmation or  off-chain payment failure
2019-05-11 12:48:24.358 [INF] CLIENT: Received block 575536
2019-05-11 12:48:24.359 [INF] CLIENT: 8f60fe Checking preimage reveal height 575565 exceeded (height 575536)
2019-05-11 12:49:09.866 [WRN] LNDCLIENT: Payment 13bad864df40a754521115300dbe9a826e214e9353982fc7415892c28c8f65a9 failed: UnknownPaymentHash(amt=27914000 mSAT)
2019-05-11 12:49:09.866 [INF] CLIENT: 8f60fe Failed prepayment: UnknownPaymentHash(amt=27914000 mSAT)
2019-05-11 12:49:09.866 [INF] CLIENT: 8f60fe Wait for server pulling off-chain payment(s)
```

The prepayment failed quickly with `Failed prepayment: UnknownPaymentHash`, but monitor showed it stuck at 'INITIATED' for days:
```
2019-05-11T12:48:12Z LOOP_OUT INITIATED 0.005 BTC - bc1qdph5mvhpsnra73mnjzvvayhxzuzf8a5f9tzy0eggcz2judrel54qgl2xhm
```

Today I restarted my Loop server, and it updated to 'FAILED'.
```
2019-05-13T10:46:09Z LOOP_OUT FAILED 0.005 BTC - bc1qdph5mvhpsnra73mnjzvvayhxzuzf8a5f9tzy0eggcz2judrel54qgl2xhm
```

The payments sent out were for: 485,707 sats and 27,914 sats, for a total of: 513,621 sats
Now that channel is being force-closed, and I don't know why. It shows a balance of 489,395sats where before it had a balance of 983,380 sats.

So far it looks like:
* The payment for 485,707 sats went out
* The payment for 27,914 sats failed (quickly) with UnknownPaymentHash
* The Loop server didn't fail the Loop Out when the second payment failed
* Restarting the Loop server kicked it in to failing the Loop Out
* Nothing has appeared in my wallet using address: bc1qdph5mvhpsnra73mnjzvvayhxzuzf8a5f9tzy0eggcz2judrel54qgl2xhm
* Something is force-closing that channel

My main question is: what has happened to the 485,707 sats payment that went out? Will it time out or come back? Or is it gone forever? If it times out, how will it come back to a channel that's forced-closing or forced-closed?

My second question is: did Loop cause the force-closing of the channel?

My third question: can we get Loop to handle the initial 'Failed prepayment: UnknownPaymentHash' better?

This has happened several times (including the force-close) but I've just used the logs for one instance. I can provide others if it'll help.

I wish I was better at tracing what has happened. I have tried. If you've any tips or advice that can show me how to trace what has happened, please do let me know.

**Comment by alexbosworth** [comment]
The UnknownPaymentHash error is a known issue that can happen if blocks are mined too quickly during the period where a payment is making its way to the service. As a workaround you can try initiating another swap.

If you consider payments off-chain, they are regular off-chain payments and can have a timeout length depending on the route your node chose. Since they are regular off-chain payments, there is nothing different about them with regard to force closes, other than the fact that they may be pending longer than usual.

Potentially we should allow directing a maximum value for this in the swap command. In a future Loop release we can also reduce the pending time of the swap.

**Comment by OpinionatedGeek** [comment]
> The UnknownPaymentHash error is a known issue that can happen if blocks are mined too quickly during the period where a payment is making its way to the service. As a workaround you can try initiating another swap.

Thanks. Is this after the UnknownPaymentHash error but before the Loop server recognises it as 'FAILED'? Or is it possible to use this to fix up the current situation with the force-closed channels?

> If you consider payments off-chain, they are regular off-chain payments and can have a timeout length depending on the route your node chose. Since they are regular off-chain payments, there is nothing different about them with regard to force closes, other than the fact that they may be pending longer than usual.

I don't think I follow this. If a payment went out but is still unsettled, it will time out. I get that. But I'm lost when it comes to how that would time out when the channel is force-closing. In a regular situation the funds are locked from the channel while the payment is in-flight (I think), and if the payment times out those funds are unlocked. But where do the funds come back to in the force-closed situation? Should I expect the channel closure to return the full balance to my wallet (including the timed-out payment)?

Also - how can I inspect these in-flight payments? They don't appear in `listpayments`. I'd like to find out when I can expect them to time out.

Many thanks,

Geoff

**Comment by OpinionatedGeek** [comment]
Some further information on the force close now that it has completed. Here are the close details from RTL:
```
Close Type: LOCAL_FORCE_CLOSE 
Channel Point: 489d2ef2b904ec0b0687517ad52451f868987f13b2073b5cc1b807c3f0d61c95:0 
Chan Id: 632050960802447360 
Closing Tx Hash: 2560a2338ca07f12d74bdf31c37aa5ca357892051a61f50266fac21e891e89b4 
Remote Pubkey: 03bb88ccc444534da7b5b64b4f7b15e1eccb18e102db0e400d4b9cfe93763aa26d 
Capacity: 1,000,000
Close Height: 575,820
Settled Balance: 489,395
Time Locked Balance: 956,686
```
Here's a link to the [closing transaction](https://live.blockcypher.com/btc/tx/2560a2338ca07f12d74bdf31c37aa5ca357892051a61f50266fac21e891e89b4/)

It looks like the 'swap payment' is still in-flight - how can I check when it will time out, since it doesn't appear in `listpayments`? And what will happen when it times out and the channel is closed?

**Comment by bjornoj** [comment]
> It looks like the 'swap payment' is still in-flight - how can I check when it will time out, since it doesn't appear in `listpayments`? And what will happen when it times out and the channel is closed?

OPTIONS:
   --include_incomplete  if set to true, payments still in flight (or failed) will be returned as well

So just do
```lncli listpayments --include_incomplete```

**Comment by alexbosworth** [comment]
If you are using the API, there is also the option to lookup a payment by its id using the track payment router rpc api

**Comment by joostjager** [comment]
I tried to reproduce this issue, but didn't succeed.

If I simulate:
- swap payment accepted by server
- prepay payment failed
- prepay invoice times out on the server after a short period (default: 180 seconds)
- server cancels swap invoice
- client reports swap finished in state `FailOffchainPayments`.

**Comment by arshbot** [comment]
I can reproduce the issue.

Logs After attempting swap:

```
2019-08-01 13:07:13.124 [INF] CLIENT: Offchain swap destination: 0223acffd7f363b4591ce860eda870fea352e981212d8a25e96a0ebea37faae288                                                          
2019-08-01 13:07:24.853 [INF] LOOPD: Loop out request received
2019-08-01 13:07:24.859 [INF] CLIENT: LoopOut 0.005 BTC to tb1q34tkz0xsgnzrn37skzzl8eeu8qy935vchp0f0y (channel: 0xc00037e270)                                                                
2019-08-01 13:07:24.865 [INF] CLIENT: Initiating swap request at height 1571924
2019-08-01 13:07:25.186 [INF] CLIENT: e7f80c Htlc address: tb1q2upf3zn682kp5f7d6l2ghkkzz4euq6zrn5ma2u0hj44rhhf34kmqkk2x08                                                                    
2019-08-01 13:07:25.191 [INF] CLIENT: e7f80c state Initiated
2019-08-01 13:07:25.191 [INF] CLIENT: e7f80c Sending swap payment lntb5000100n1pw59eqdpp5uluqecrw3stes390xzsqrzpku9fmxg5kgmv2wjej6at04ww2ev8qdq8wdmkzuqcqzxgxq97zvuq4zrhzc88vuqvvqdaltdm6py8mp4kpxfs432jgu84pnne40ze5ttzl6hfxyl07hhvtezcdsr2env938k3xz73c8yr3j276vzg9mfv67gpm845g5
2019-08-01 13:07:25.191 [INF] CLIENT: e7f80c Sending prepayment lntb13370n1pw59eqdpp52j0q2yr7qjvwcxvawh49tkyxejjgx8v576xsccxvmh445t0sej2sdq2wpex2urp0ycqzxgxq97zvuqcq9xxm3s56djekhh0fj8hp8ff997galgsr95w5c56dsw40je6fznssj7mh2j9zldcjkce27rt03mpkyt3wqsdnn7fxllck94kchmnaqpmew259
2019-08-01 13:07:25.192 [INF] CLIENT: e7f80c Register conf ntfn for swap script on chain (hh=1571924)                                                                                        
2019-08-01 13:07:25.193 [INF] CLIENT: e7f80c Checking preimage reveal height 1572004 exceeded (height 1571924)                                                                               
2019-08-01 13:07:25.193 [INF] CLIENT: e7f80c Waiting for either htlc on-chain confirmation or  off-chain payment failure                                                                     
2019-08-01 13:07:42.594 [INF] LNDCLIENT: Payment 549e05107e0498ec199d75ea55d886cca4831d94f68d0c60ccddeb5a2df0cc95 completed                                                                  
2019-08-01 13:07:51.550 [WRN] LNDCLIENT: Payment f11ba1b44a0303e5758553fabad1dbe1bdece68d397df292f7dc944daa8967d5 failed: UnknownPaymentHash(amt=1337000 mSAT)                               
2019-08-01 13:07:51.550 [INF] CLIENT: 48c994 Prepayment failed: UnknownPaymentHash(amt=1337000 mSAT)                                                                                         
2019-08-01 13:07:51.550 [INF] CLIENT: 48c994 Swap completed: FailOffchainPayments (final cost: server 0 BTC, onchain 0 BTC, offchain 0 BTC)                                                  
2019-08-01 13:07:51.553 [INF] CLIENT: 48c994 state FailOffchainPayments
```

Loop monitor output:

```
2019-08-01T13:07:25Z LOOP_OUT INITIATED 0.005 BTC - tb1q2upf3zn682kp5f7d6l2ghkkzz4euq6zrn5ma2u0hj44rhhf34kmqkk2x08                                                                           
2019-08-01T13:07:51Z LOOP_OUT FAILED 0.005 BTC - tb1qssmhwpzll9yz6awzq57rh0qyyd9ykajmk54qjsnup6ygdjznxtzs5xqey5 (cost: server 0, onchain 0, offchain 0) 
```

Queryroutes output: 

```

**Comment by joostjager** [comment]
@arshbot what I tried to reproduce is the swap not moving to the FAILED final state after the prepay fails. I didn't succeed in that and it seems that in your case it also moves properly to the FAILED state.

The other question is why you are getting the unknown payment hash for the prepay. I checked the server logs and it shows that the prepay payment was accepted, but the swap payment didn't make it to the server in time (10 mins). When that happens the prepay payment is released and returns to the client as an unknown payment hash.

You'll need your logs from a bit further back in time, around 2019-08-01 12:57:33, to see the swap payment being initiated.
